wolfcrypt/src/ed25519.c

```diff
@@ -56,30 +56,47 @@ static const byte ed25519Ctx[ED25519CTX_SIZE+1] =
                                              "SigEd25519 no Ed25519 collisions";
 #endif
 
+static int ed25519_hash_reset(ed25519_key* key) {
+    int ret;
+    if (key->sha_clean_flag)
+        ret = 0;
+    else {
+        wc_Sha512Free(&key->sha);
+        ret = wc_InitSha512_ex(&key->sha, key->heap, INVALID_DEVID);
+        if (ret == 0)
+            key->sha_clean_flag = 1;
+    }
+    return ret;
+}
+
+static int ed25519_hash_update(ed25519_key* key, const byte* data, word32 len) {
+    if (key->sha_clean_flag)
+        key->sha_clean_flag = 0;
+    return wc_Sha512Update(&key->sha, data, len);
+}
+
+static int ed25519_hash_final(ed25519_key* key, byte* hash) {
+    int ret = wc_Sha512Final(&key->sha, hash);
+    if (ret == 0)
+        key->sha_clean_flag = 1;
+    return ret;
+}
+
 static int ed25519_hash(ed25519_key* key, const byte* in, word32 inLen,
     byte* hash)
 {
     int ret;
-    wc_Sha512 sha;
-    int devId = INVALID_DEVID;
 
     if (key == NULL || (in == NULL && inLen > 0) || hash == NULL) {
         return BAD_FUNC_ARG;
     }
 
-#ifdef WOLF_CRYPTO_CB
-    devId = key->devId;
-#endif
-
-    ret = wc_InitSha512_ex(&sha, NULL, devId);
-    if (ret == 0) {
-        ret = wc_Sha512Update(&sha, in, inLen);
-        if (ret == 0)
-            ret = wc_Sha512Final(&sha, hash);
-        wc_Sha512Free(&sha);
-    }
+    ret = ed25519_hash_reset(key);
```

Comment:
There doesn't appear to be a need for this change.

Reply:
This is required to assure that the sha is reset to a known clean state before using it.  In the previous code, a fresh sha object was allocated, initialized, and deallocated for every call to `ed25519_hash()` and the other routines that computed hashes.  With the new code, the sha state is persistent, and `.sha_clean_flag` allows `ed25519_hash_reset()` to know whether a deallocation-initialization cycle is needed.

In the absence of runtime and/or usage errors, `ed25519_hash_reset()` is a no-op (a meaningful optimization), but I did not want to introduce a new path to undefined behavior.  Therefore `ed25519_hash_reset()` precisely replicates the old behavior when necessary.